src/graph.py

```diff
@@ -0,0 +1,95 @@
+"""Implement a class for a graph data structure."""
+
+
+class Graph(object):
+    """Graph class."""
+
+    def __init__(self):
+        """Initialize a graph."""
+        self._graph = {}
+
+    def nodes(self):
+        """Return a list of all nodes in graph."""
+        return list(self._graph.keys())
+
+    def edges(self):
+        """Return a list of edges in graph."""
+        edges = []
+        for key in self._graph:
+            for i in self._graph[key]:
+                edges.append((key, i))
+        return edges
+
+    def add_node(self, val):
+        """Add a node with value of val to graph."""
+        self._graph[val] = {}
+
+    def add_edge(self, val1, val2, weight=0):
+        """Add a new edge to graph between val1 & val2 as well as add vals."""
+        if val1 in self._graph and val2 in self._graph:
+            if val2 not in self._graph[val1]:
+                self._graph[val1][val2] = weight
+        elif val1 in self._graph and val2 not in self._graph:
+            self._graph[val2] = {}
+            self._graph[val1][val2] = weight
+        elif val2 in self._graph and val1 not in self._graph:
+            self._graph[val1] = {}
+            self._graph[val1][val2] = weight
+        else:
+            self._graph[val1] = {val2: weight}
+            self._graph[val2] = {}
```

Comment:
You could refactor this whole set of `if` statements to ask questions and do operations like...
```
is val1 in the graph? if not, add it
is val2 in the graph? if not, add it
is val2 one of val1's neighbors? if not, add it as a neighbor with the given weight. If so, re-assign the weight of that edge to be the given weight.
```